packages/salesforcedx-vscode-soql/src/queryDataView/queryDataFileService.ts

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import { JsonMap } from '@salesforce/ts-types';
+import * as fs from 'fs';
+import { QueryResult } from 'jsforce';
+import * as path from 'path';
+import * as vscode from 'vscode';
+import { QUERY_RESULTS_DIR_PATH } from '../constants';
+import {
+  CsvDataProvider,
+  DataProvider,
+  JsonDataProvider
+} from './dataProviders';
+
+export enum FileFormat {
+  JSON = 'json',
+  CSV = 'csv'
+}
+
+const sfdxCoreExtension = vscode.extensions.getExtension(
+  'salesforce.salesforcedx-vscode-core'
+);
+const sfdxCoreExports = sfdxCoreExtension
+  ? sfdxCoreExtension.exports
+  : undefined;
+const { getRootWorkspacePath } = sfdxCoreExports;
+
+export class QueryDataFileService {
+  private dataProvider: DataProvider;
+
+  constructor(
+    private queryData: QueryResult<JsonMap>,
+    private format: FileFormat,
+    private documentName: string
+  ) {
+    this.dataProvider = this.getDataProvider();
+  }
+
+  protected getDataProvider(): DataProvider {
+    switch (this.format) {
+      case FileFormat.CSV:
+        return new CsvDataProvider(this.documentName);
+      case FileFormat.JSON:
+        return new JsonDataProvider(this.documentName);
+      default:
+        throw new Error('No DataProvider Found');
+    }
+  }
+
+  public save(): string {
+    try {
+      const fileContent = this.dataProvider.getFileContent(
+        this.queryData.records
+      );
+      const savedFileName = this.dataProvider.getFileName();
+      const queryDataFilePath = path.join(
+        this.getResultsDirectoryPath(),
+        savedFileName
+      );
+
+      this.createResultsDirectoryIfDoesNotExist();
+      // Save query results to disk
+      fs.writeFileSync(queryDataFilePath, fileContent);
+      this.showSaveSuccessMessage(savedFileName);
+      this.showFileInExplorer(queryDataFilePath);
+
+      return queryDataFilePath;
+    } catch (error) {
+      // TODO: i18n, CCX
+      vscode.window.showErrorMessage(
+        `Your data could not be saved. Run the query and try again.`
```

Comment:
@ekapner This message will be shown to the users if something goes wrong either creating the directory to store files or converting the json data to csv or Json data to a string.
Either way there is not much the customer can do about it except verify there is no error or something weird with the SOQL query itself or so API change/issue.

Reply:
How about this, @jonnyhork? 

We couldn't save your data. Verify the query syntax, then run the query again.

Reply:
Sounds good to me

Reply:
Per our convo in Slack:

We couldn't save your results. Verify the query syntax, then run the query again.

Reply:
I'm thinking "the" sounds better than "your"

Reply:
okey dokey